File: java/28.java
// 28. Implement strStr()
// Implement strStr().
//
// Return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
//
// Clarification:
//
// What should we return when needle is an empty string? This is a great question to ask during an interview.
//
// For the purpose of this problem, we will return 0 when needle is an empty string. This is consistent to C's strstr() and Java's indexOf().
//
//
//
// Example 1:
//
// Input: haystack = "hello", needle = "ll"
// Output: 2
// Example 2:
//
// Input: haystack = "aaaaa", needle = "bba"
// Output: -1
// Example 3:
//
// Input: haystack = "", needle = ""
// Output: 0
//
//
// Constraints:
//
// 0 <= haystack.length, needle.length <= 5 * 104
// haystack and needle consist of only lower-case English characters.
//
// Runtime: 1697 ms, faster than 9.26% of Java online submissions for Implement strStr().
// Memory Usage: 39 MB, less than 37.34% of Java online submissions for Implement strStr().
class Solution {
    public int strStr(String haystack, String needle) {
        if (needle.equals("")) return 0;
        if (needle.length() > haystack.length()) return -1;
        int start = -1;
        for (int i = 0; i < haystack.length(); i++) {
            for (int j = 0; j < needle.length(); j++) {
                if (i + j == haystack.length()) return -1;
                if (needle.charAt(j) != haystack.charAt(i + j)) break;
                if (j == needle.length() - 1) return i;
            }
        }
        return -1;
    }
}

// Runtime: 308 ms, faster than 28.91% of Java online submissions for Implement strStr().
// Memory Usage: 39.7 MB, less than 7.97% of Java online submissions for Implement strStr().
public class Solution {
    public int strStr(String haystack, String needle) {
        if (needle.equals("")) return 0;
        if (needle.length() > haystack.length()) return -1;
        int len = needle.length();
        for (int i = 0; i <= haystack.length() - len; ++i) {
            if (haystack.substring(i, i + len).equals(needle)) {
                return i;
            }
        }
        return -1;
    }
}

File: go/28.go
// 28. Find the Index of the First Occurrence in a String

// Given two strings needle and haystack, return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.

 

// Example 1:

// Input: haystack = "sadbutsad", needle = "sad"
// Output: 0
// Explanation: "sad" occurs at index 0 and 6.
// The first occurrence is at index 0, so we return 0.
// Example 2:

// Input: haystack = "leetcode", needle = "leeto"
// Output: -1
// Explanation: "leeto" did not occur in "leetcode", so we return -1.
 

// Constraints:

// 1 <= haystack.length, needle.length <= 104
// haystack and needle consist of only lowercase English characters.

// Runtime 0 ms Beats 100.00%
// Memory 3.99 MB Beats 49.47%
func strStr(haystack string, needle string) int {
    if len(haystack) < len(needle) {
        return -1
    }
    needleIndex := 0
    init := 0
    for i := 0; i < len(haystack); i++ {
        if haystack[i] != needle[needleIndex] {
            if needleIndex != 0 {
                i = init
            }
            needleIndex = 0
        } else {
            if needleIndex == 0 {
                init = i
            }
            needleIndex++
            if needleIndex == len(needle) {
                return init
            } 
        }
    }
    return -1
}


// Runtime 0 ms Beats 100.00%
// Memory 3.95 MB Beats 49.47%
func strStr(haystack string, needle string) int {
    if len(haystack) < len(needle) {
        return -1
    }
    for i := 0; i < len(haystack); i++ {
        if haystack[i] == needle[0] && i + len(needle) <= len(haystack) {
            if haystack[i: (i + len(needle))] == needle {
                return i
            }
        }
    }
    return -1
}